### PR TITLE
Pp 5302 search by state and dates

### PR DESCRIPTION
**Pull request by oswaldquek**
## WHAT YOU DID
_A brief description of the pull request:_

## How to test

- How should it be reviewed? 
- What feedback do you need? 
- If manual testing is needed, give suggested testing steps.

### Logging

- [ ] only emit log lines at ERROR level which require immediate attention from a support engineer. These will trigger a zendesk alert.

### Documentation

- [ ] Updated README.md for any of the following ?

* Introduced any new environment variables / removed existing environment variable
* Added new API / updated existing API definition

### CODE REVIEW

File: src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java [hunk 1]
```diff
@@ -86,13 +88,19 @@ public MandateSearchDao(Jdbi jdbi) {
...
+        params.getGreaterThanSearchTermsOnMandate().forEach(e -> queryMap.put(e.getKey() + "_from", e.getValue()));
+        params.getLessThanSearchTermsOnMandate().forEach(e -> queryMap.put(e.getKey() + "_to", e.getValue()));
         return queryMap;
     }
 
     private String generateQuery(Params params) {
         StringBuilder sb = new StringBuilder();
         params.getCaseInsensitivePartialSearchTermsOnMandate().forEach(e -> sb.append(format(" AND m.%s ILIKE :%s", e.getKey(), e.getKey())));
         params.getCaseInsensitivePartialSearchTermsOnPayer().forEach(e -> sb.append(format(" AND p.%s ILIKE :%s", e.getKey(), e.getKey())));
+        params.getExactSearchTermsOnMandate().forEach(e -> sb.append(format(" AND m.%s = :%s", e.getKey(), e.getKey())));
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java]: If we’ve already got a `StringBuilder`, we don’t really need to use `String.format(…)`, we can just do something like this:

```java
sb.append(" AND m.").append(e.getKey()).append(" = :").append(e.getKey());
```

File: src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java [hunk 2]
```diff
@@ -86,13 +88,19 @@ public MandateSearchDao(Jdbi jdbi) {
...
+        params.getGreaterThanSearchTermsOnMandate().forEach(e -> sb.append(format(" AND m.%s > :%s", e.getKey(), e.getKey() + "_from")));
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java]: It took me more time than I care to admit to determine that this doesn’t introduce a risk of SQL injection (because `e.getKey()` comes from predefined constant strings defined by us rather than any remote input. I wish this was more obvious.

File: src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java [hunk 3]
```diff
@@ -117,17 +125,29 @@ private Params(MandateSearchParams mandateSearchParams) {
...
+        List<SimpleEntry<String, ?>> getLessThanSearchTermsOnMandate() {
+            return List.of(toDate).stream().filter(e -> e.getValue() != null).collect(Collectors.toList());
+        }
+
+        List<SimpleEntry<String, ?>> getGreaterThanSearchTermsOnMandate() {
+            return List.of(fromDate).stream().filter(e -> e.getValue() != null).collect(Collectors.toList());
+        }
+
+        List<SimpleEntry<String, ?>> getExactSearchTermsOnMandate() {
+            return List.of(mandateState).stream().filter(e -> isNotBlank(e.getValue())).collect(Collectors.toList());
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java]: Why is this one `isNotBlank(…)` while the other ones are `!= null`?

File: src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java [hunk 4]
```diff
@@ -117,17 +125,29 @@ private Params(MandateSearchParams mandateSearchParams) {
...
+        }
+
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java]: I know you’re just modifying what’s already there but I find this class quite hard to follow. I’m not saying change it now, but fundamentally I don’t see why it can’t be simplified to something like this:

```java
public List<Mandate> search(MandateSearchParams mandateSearchParams) {
	var sql = new StringBuilder();
	sql.append("SELECT DISTINCT … WHERE g.external_id = :gatewayAccountExternalId");

	var sqlParams = new HashMap<>();
	sqlParams.put("gatewayAccountExternalId", mandateSearchParams.getGatewayAccountExternalId());

	Optional.ofNullable(mandateSearchParams.getReference()).ifPresent(reference -> {
		sql.append(" AND m.reference ILIKE :reference");
		sqlParams.put("reference", reference);
	});

	Optional.ofNullable(mandateSearchParams.getFromDate()).ifPresent(fromDate -> {
		sql.append(" AND m.created_date_from > :created_date_from");
		sqlParams.put("created_date_from", fromDate);
	});

	// etc. etc.

    return jdbi.withHandle(handle -> {
        Query query = handle.createQuery(sql.toString());
        sqlParams.forEach(query::bind);
        return query.map(new MandateMapper()).list();
    });
}
```

It would be more verbose but I think it would be a lot easier to follow.

File: src/test/java/uk/gov/pay/directdebit/mandate/dao/SearchMandateDaoIT.java [hunk 5]
```diff
@@ -53,59 +59,84 @@ public void setup() {
     @Test
     @Parameters({"REF1234", "ref1234", "f12"})
     public void searchByReference(String searchString) {
-        var mandateSearchParams = aMandateSearchParams().withReference(searchString).withGatewayAccountId("gateway-account-id");
-        List<Mandate> results = mandateSearchDao.search(mandateSearchParams);
+        var searchParams = aMandateSearchParams().withReference(searchString).withGatewayAccountId("gateway-account-id");
+        List<Mandate> results = mandateSearchDao.search(searchParams);
         assertThat(results.size()).isEqualTo(1);
-        assertThat(results.get(0)).isEqualTo(mandate1.toEntity());
+        assertThat(results.get(0).getId()).isEqualTo(mandate1.getId());
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/directdebit/mandate/dao/SearchMandateDaoIT.java]: Why is this now only comparing the ID rather than the entire mandate?

File: src/test/java/uk/gov/pay/directdebit/mandate/dao/SearchMandateDaoIT.java [hunk 6]
```diff
@@ -53,59 +59,84 @@ public void setup() {
...
-        
+        var searchParams = aMandateSearchParams().withFromDate(now().minusHours(1)).withGatewayAccountId("gateway-account-id");
+        List<Mandate> results = mandateSearchDao.search(searchParams);
+        assertThat(results.size()).isEqualTo(1);
+        assertThat(results.get(0).getId()).isEqualTo(mandate1.getId());
+
+        searchParams = aMandateSearchParams().withFromDate(now().minusHours(7)).withGatewayAccountId("gateway-account-id");
+        results = mandateSearchDao.search(searchParams);
+        assertThat(results.size()).isEqualTo(2);
+        assertThat(results.stream().map(m -> m.getId()).collect(toList())).contains(mandate1.getId(), mandate2.getId());
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/directdebit/mandate/dao/SearchMandateDaoIT.java]: `assertThat(results.stream().map(Mandate::getId).collect(toSet()), is(List.of(mandate1.getId(), mandate2.getId()))` might be clearer and will also check the size.

File: src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java (hunk 4 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java]: It was the verbosity I was trying to avoid! Hence my attempt to make this more elegant by returning lists and mapping over them. However I think you're right, it's getting quite hard to read and I will use the more verbose style you have suggested for readability.

File: src/test/java/uk/gov/pay/directdebit/mandate/dao/SearchMandateDaoIT.java (hunk 5 above)
Comment (oswaldquek) [reply src/test/java/uk/gov/pay/directdebit/mandate/dao/SearchMandateDaoIT.java]: For some reason the date returned from the postgres database was BST, while it was UTC when setting up the mandate fixture. This caused the object equality to fail. I'm not sure if this is a quirk of the postgres-in-docker database.

Comment (oswaldquek) [reply src/test/java/uk/gov/pay/directdebit/mandate/dao/SearchMandateDaoIT.java]: it was actually me not setting the UTC timezone when inserting the mandate into the database.
